### PR TITLE
Avoid unnecessary copies of Payload data when not using ipc channel

### PR DESCRIPTION
**Pull request by jrmuizel**
This improves the performance of
http://people.mozilla.com/~jmuizelaar/implementation-tests/dl-test.yaml
in wrench from 17fps to 19.5fps on my machine

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/1041)

### CODE REVIEW

**Comment by jrmuizel** [comment]
This is still a bit messy, but I wanted to get feedback early.

**Comment by jrmuizel** [comment]
This changes the serialization of Payload to include the length of the two parts. I'll bump the binary recording version.

File: webrender_traits/src/channel.rs [hunk 1]
```diff
@@ -2,10 +2,52 @@
...
+#[derive(Clone)]
+pub struct Payload {
+    pub epoch: Epoch,
+    pub display_list_data: Vec<u8>,
+    pub auxiliary_lists_data: Vec<u8>
+}
+
+impl Payload {
+    pub fn to_data(&self) -> Vec<u8> {
+        let mut data = vec![];
```
Comment (kvark) [inline webrender_traits/src/channel.rs]: we could use `with_capacity`

File: webrender_traits/src/channel.rs [hunk 2]
```diff
@@ -2,10 +2,52 @@
...
+
+#[derive(Clone)]
+pub struct Payload {
+    pub epoch: Epoch,
+    pub display_list_data: Vec<u8>,
+    pub auxiliary_lists_data: Vec<u8>
+}
+
+impl Payload {
+    pub fn to_data(&self) -> Vec<u8> {
```
Comment (kvark) [inline webrender_traits/src/channel.rs]: nit: consider using `From`/`Into` traits. Not really helping anything, just seems to be more idiomatic

File: webrender_traits/src/channel_ipc.rs [hunk 3]
```diff
@@ -19,9 +21,16 @@ pub type PayloadSender = IpcBytesSender;
 
 pub type PayloadReceiver = IpcBytesReceiver;
 
-impl PayloadHelperMethods for PayloadSender {
-    fn send_vec(&self, data: Vec<u8>) -> Result<(), Error> {
-        self.send(&data)
+impl PayloadSenderHelperMethods for PayloadSender {
+    fn send(&self, data: Payload) -> Result<(), Error> {
+        self.send(&data.to_data())
```
Comment (kvark) [inline webrender_traits/src/channel_ipc.rs]: is there any benefit of heap-allocating the vector, serializing the data to it, and then sending, versus just sending the data right away during serialization?

File: webrender_traits/src/api.rs [hunk 4]
```diff
@@ -305,20 +304,18 @@ impl RenderApi {
...
+                                             aux_desc,
                                              preserve_frame_state);
         self.api_sender.send(msg).unwrap();
 
-        let mut payload = vec![];
-        payload.write_u32::<LittleEndian>(epoch.0).unwrap();
-        payload.extend_from_slice(display_list.data());
-        payload.extend_from_slice(auxiliary_lists.data());
-        self.payload_sender.send_vec(payload).unwrap();
+        PayloadSenderHelperMethods::send(&self.payload_sender, Payload{epoch: epoch, display_list_data: dl_data, auxiliary_lists_data: aux_data}).unwrap();
```
Comment (kvark) [inline webrender_traits/src/api.rs]: what's the reason of not calling it like `self.payload_sender.send(...)`?

File: webrender/src/render_backend.rs [hunk 5]
```diff
@@ -188,36 +186,26 @@ impl RenderBackend {
                             let mut leftover_auxiliary_data = vec![];
                             let mut auxiliary_data;
                             loop {
-                                auxiliary_data = self.payload_rx.recv().unwrap();
+                                auxiliary_data = PayloadReceiverHelperMethods::recv(&self.payload_rx).unwrap();
```
Comment (kvark) [inline webrender/src/render_backend.rs]: similarly, what's the reason of not doing `self.payload_rx.recv()`?

File: webrender_traits/src/channel_ipc.rs (hunk 3 above)
Comment (jrmuizel) [inline webrender_traits/src/channel_ipc.rs]: Yes. send(Vec) is still message based. So the receiver expects to get everything in recv call. In the long term we'd probably want to fix this to be more efficient, but it looks like ipc-channel has lower hanging efficiency problems to deal with first.

**Comment by kvark** [comment]
Thanks!
@bors-servo r+

**Comment by kvark** [comment]
Classy timout on OSX
@bors-servo retry

**Pull request merged**